### PR TITLE
PP601: Port the GasEstimator from the server to the client

### PR DESCRIPTION
**Pull request by AndresQuijano**
## What

- Port the GasEstimator from rif-relay-server to rif-relay-client

## Why

- Project refactoring

## Refs

- [PP-601](https://rsklabs.atlassian.net/browse/PP-601)

### CODE REVIEW

**Comment by franciscotobar** [comment]
Need to merge https://github.com/rsksmart/rif-relay-client/pull/60 and rebase.

File: src/gasEstimator/gasEstimator.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+): Promise<BigNumber> => {
+  const {
+    relayRequest,
+    metadata: { signature },
+  } = request;
+
+  const { tokenContract, tokenAmount } = relayRequest.request;
+  const { feesReceiver, callForwarder, gasPrice } = relayRequest.relayData;
+
+  const relayClient = new RelayClient();
```
Comment (franciscotobar) [inline src/gasEstimator/gasEstimator.ts:22]: By doing this here I think that we are creating a dependency on the relayClient and the constructor change, we will need to include those changes here or even pass the client as parameter.

File: src/gasEstimator/utils.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  if (isDeployRequest(envelopingRequest)) {
+    throw Error('LinearFit estimation not implemented for deployments');
+  }
+
+  const {
+    request: { data, from, to },
+    relayData: { gasPrice },
+  } = envelopingRequest;
+
+  const relayClient = new RelayClient();
```
Comment (franciscotobar) [inline src/gasEstimator/utils.ts:64]: By doing this here I think that we are creating a dependency on the relayClient and the constructor change, we will need to include those changes here or even pass the client as parameter.

File: src/gasEstimator/gasEstimator.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+  const isSmartWalletDeploy = isDeployRequest(relayRequest);
+
+  //FIXME validate how to generate the smart wallet address
+  const preDeploySWAddress = isSmartWalletDeploy ? undefined : undefined;
```
Comment (franciscotobar) [inline src/gasEstimator/gasEstimator.ts:27]: To generate the SW Address I could generate my own function or use https://github.com/rsksmart/rif-relay-client/blob/PP601_gasEstimator/src/discovery/utils.ts#L147, but this adds more complexity for a small task.

File: src/RelayClient.ts [hunk 4]
```diff
@@ -53,19 +56,33 @@ type RequestConfig = {
   estimatedGasCorrectionFactor?: BigNumberish;
 };
 
+//FIXME name standardization
 type TokenGasEstimationParams = Pick<
```
Comment (antomor) [inline src/RelayClient.ts:60]: IMHO having the types defined in this way makes the code very difficult to read.
No need to change it now, but we may probably discuss it with the rest of the team

File: src/RelayClient.ts [hunk 5]
```diff
@@ -53,19 +56,33 @@ type RequestConfig = {
...
+  Pick<
+    RequestConfig,
+    | 'isSmartWalletDeploy'
+    | 'preDeploySWAddress'
+    | 'internalEstimationCorrection'
+    | 'estimatedGasCorrectionFactor'
+  >;
+
+//FIXME name standardization
 type EstimateInternalGasParams = Pick<
```
Comment (antomor) [inline src/RelayClient.ts:77]: IMHO having the types defined in this way makes the code very difficult to read.
No need to change it now, but we may probably discuss it with the rest of the team

File: src/gasEstimator/gasEstimator.ts (hunk 1 above)
Comment (antomor) [reply src/gasEstimator/gasEstimator.ts:22]: Does this mean we're going to have a circular dependency?

File: src/gasEstimator/gasEstimator.ts (hunk 3 above)
Comment (antomor) [reply src/gasEstimator/gasEstimator.ts:27]: totally agree, we may consider this improvement in future.

File: src/gasEstimator/utils.ts (hunk 2 above)
Comment (antomor) [reply src/gasEstimator/utils.ts:64]: does this mean we're going to introduce a circular dependency?

**Review by antomor — APPROVED** [review]
@franciscotobar thank you for the changes, I've just left doubts, no changes strictly required.

File: src/gasEstimator/gasEstimator.ts (hunk 1 above)
Comment (franciscotobar) [reply src/gasEstimator/gasEstimator.ts:22]: not circular but we are coupling this method with the relayClient.

File: src/gasEstimator/utils.ts (hunk 2 above)
Comment (franciscotobar) [reply src/gasEstimator/utils.ts:64]: not circular but we are coupling this method with the relayClient.

Comment (antomor) [reply src/gasEstimator/utils.ts:64]: I think we can simplify the interaction later on. I'd leave it as it is for now

**Pull request merged**